server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ClusterBalanceStats.java

```diff
@@ -50,28 +53,55 @@ public static ClusterBalanceStats createFrom(
             if (dataRoles.isEmpty()) {
                 continue;
             }
-            var nodeStats = NodeBalanceStats.createFrom(routingNode, clusterState.metadata(), clusterInfo, writeLoadForecaster);
+            var nodeStats = NodeBalanceStats.createFrom(
+                routingNode,
+                clusterState.metadata(),
+                desiredBalance,
+                clusterInfo,
+                writeLoadForecaster
+            );
             nodes.put(routingNode.node().getName(), nodeStats);
             for (DiscoveryNodeRole role : dataRoles) {
                 tierToNodeStats.computeIfAbsent(role.roleName(), ignored -> new ArrayList<>()).add(nodeStats);
             }
         }
-        return new ClusterBalanceStats(Maps.transformValues(tierToNodeStats, TierBalanceStats::createFrom), nodes);
+        return new ClusterBalanceStats(
+            nodes.values().stream().mapToInt(NodeBalanceStats::shards).sum(),
```

Comment:
Consider iterating over nodes only once to accumulate the stats

```java
var shardsUndesiredShardsAllocation = nodes.values()
            .stream()
            .collect(
                Collectors.teeing(
                    Collectors.summingInt(NodeBalanceStats::shards),
                    Collectors.summingInt(NodeBalanceStats::undesiredShardAllocations),
                    Tuple::tuple
                )
            );
```

Reply:
TIL: teeing collector,

however I would like to stick with existing approach as collection is small (same as amount of nodes in the cluster) and does not contain a lot of filter/map operations on the way